price-estimator/src/solver_rounding_buffer.rs

```diff
@@ -0,0 +1,162 @@
+
+    // Reduce account balances.
+    for ((address, token_id), rounding_buffer) in account_balance_buffers {
+        if let Some(balance) = account_state.0.get_mut(&(address, token_id.0)) {
+            *balance = balance.saturating_sub(U256::from(rounding_buffer));
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ethcontract::{Address, U256};
+
+    fn address(n: u64) -> Address {
+        Address::from_low_u64_le(n)
+    }
+
+    fn account(address_: u64, token: u16, balance: u128) -> ((Address, u16), U256) {
+        ((address(address_), token), U256::from(balance))
+    }
+
+    fn token_info(external_price: u128) -> TokenInfo {
+        TokenInfo {
+            alias: None,
+            decimals: None,
+            external_price,
+        }
+    }
+
+    fn order(
+        id: u16,
+        address_: u64,
+        buy_token: u16,
+        sell_token: u16,
+        numerator: u128,
+        denominator: u128,
+    ) -> Order {
+        Order {
+            id,
+            account_id: address(address_),
+            buy_token,
+            sell_token,
+            numerator,
+            denominator,
+            remaining_sell_amount: denominator,
+            valid_from: 0,
+            valid_until: 0,
+        }
+    }
+
+    #[test]
+    fn apply_rounding_buffer_ok() {
+        let mut tokens = Tokens::new();
+        tokens.insert(TokenId(0), token_info(1));
+        tokens.insert(TokenId(1), token_info(2));
+        tokens.insert(TokenId(2), token_info(10));
+
+        let accounts = vec![
+            account(0, 0, 100_000_000_000_000_000),
+            account(0, 1, 100_000_000_000_000_000),
+            account(0, 2, 100_000_000_000_000_000),
+        ];
+        let mut account_state = AccountState(accounts.into_iter().collect());
+
+        let mut orders = vec![
+            order(0, 0, 1, 0, 600_000_000_000_000, 500_000_000_000_000),
+            order(1, 0, 0, 1, 600_000_000_000_000, 500_000_000_000_000),
+            order(2, 0, 2, 0, 600_000_000_000_000, 500_000_000_000_000),
+            order(3, 0, 0, 2, 600_000_000_000_000, 500_000_000_000_000),
+            order(4, 0, 2, 1, 600_000_000_000_000, 500_000_000_000_000),
+            order(5, 0, 1, 2, 600_000_000_000_000, 500_000_000_000_000),
+        ];
+
+        apply_rounding_buffer(&tokens, &mut orders, &mut account_state);
+
+        let expected_orders = vec![
+            order(0, 0, 1, 0, 600_000_000_000_000, 490_000_000_000_000),
+            order(1, 0, 0, 1, 600_000_000_000_000, 495_000_000_000_000),
+            order(2, 0, 2, 0, 600_000_000_000_000, 490_000_000_000_000),
+            order(3, 0, 0, 2, 600_000_000_000_000, 499_000_000_000_000),
+            order(4, 0, 2, 1, 600_000_000_000_000, 495_000_000_000_000),
+            order(5, 0, 1, 2, 600_000_000_000_000, 499_000_000_000_000),
+        ];
+        for (order, expected) in orders.iter().zip(expected_orders.iter()) {
+            assert_eq!(order, expected);
+        }
```

Comment:
Does equality on the vec not hold?

Reply:
Equality is the same but this gives better output in case it doesn't match because the output is reduced to the one failing element. It was useful when I wrote the test so I kept it.